python/hail/matrixtable.py

```diff
@@ -2116,27 +2126,67 @@ def index_rows(self, *exprs: Tuple[Expression]) -> StructExpression:
                 return self.rows().index(*exprs)
 
     def index_cols(self, *exprs: Tuple[Expression]) -> StructExpression:
-        exprs = [to_expr(e) for e in exprs]
-        indices, aggregations, joins = unify_all(*exprs)
-        src = indices.source
-
-        if aggregations:
-            raise ExpressionException('Cannot join using an aggregated field')
-        uid = Env.get_uid()
-
-        if src is None:
-            raise ExpressionException('Cannot index with a scalar expression')
-
         return self.cols().index(*exprs)
 
     def index_entries(self, row_exprs: Tuple[Expression], col_exprs: Tuple[Expression]) -> StructExpression:
-        row_exprs = [to_expr(e) for e in row_exprs]
-        col_exprs = [to_expr(e) for e in col_exprs]
+        row_exprs = tuple(row_exprs)
+        col_exprs = tuple(col_exprs)
+        if len(row_exprs) == 0  or len(col_exprs) == 0:
+            raise ValueError("'MatrixTable.index_entries:' 'row_exprs' and 'col_exprs' may not be empty")
+        row_non_exprs = list(filter(lambda e: not isinstance(e, Expression), row_exprs))
+        if row_non_exprs:
+            raise TypeError(f"'MatrixTable.index_entries': row_exprs expects expressions, found {row_non_exprs}")
+        col_non_exprs = list(filter(lambda e: not isinstance(e, Expression), col_exprs))
+        if col_non_exprs:
+            raise TypeError(f"'MatrixTable.index_entries': col_exprs expects expressions, found {col_non_exprs}")
+
+        def types_match(left, right):
+            return (len(left) == len(right)
+                    and all(map(lambda lr: lr[0].dtype == lr[1].dtype, zip(left, right))))
+
+        if not types_match(self.row_key.values(), row_exprs):
+            if (len(row_exprs) == 1
+                    and isinstance(row_exprs[0], TupleExpression)
+                    and types_match(self.row_key.values(), row_exprs[0])):
+                return self.index_entries(tuple(row_exprs[0]), col_exprs)
+            elif (len(row_exprs) == 1
+                  and isinstance(row_exprs[0], StructExpression)
+                  and types_match(self.row_key.values(), row_exprs[0].values())):
```

Comment:
What does `values()` do on an expression `row_exprs[0].values()`?

Reply:
StructExpression extends `Mapping[str, Expression]`. Values are the values of the mapping.